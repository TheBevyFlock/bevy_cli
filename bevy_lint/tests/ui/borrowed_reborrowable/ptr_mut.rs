//! This tests the `borrowed_reborrowable` lint, specifically when triggered on the `PtrMut` type.

#![feature(register_tool)]
#![register_tool(bevy)]
#![deny(bevy::borrowed_reborrowable)]

use bevy::ecs::ptr::{Ptr, PtrMut};
use bevy::prelude::*;

// OK: Lint does not apply to immutable references
fn immutable_reference(_param: &PtrMut) {
    // ...
}

//~| HELP: use `PtrMut` instead
//~v ERROR: parameter takes `&mut PtrMut` instead of a re-borrowed `PtrMut`
fn mutable_reference(_param: &mut PtrMut) {
    // ...
}

//~| HELP: use `PtrMut` instead
//~v ERROR: parameter takes `&mut PtrMut` instead of a re-borrowed `PtrMut`
fn mutable_reference_return<'a>(_param: &'a mut PtrMut) -> usize {
    123
}

// OK: Lint does not apply when return type relies on reference lifetime
fn mutable_reference_bounded_return<'a>(param: &'a mut PtrMut) -> Ptr<'a> {
    param.as_ref()
}

// OK: Lint does not apply when return type relies on reference lifetime
fn mutable_reference_bounded_return_complex<'a>(param: &'a mut PtrMut) -> Vec<(usize, Ptr<'a>)> {
    vec![(1, param.as_ref())]
}

fn main() {
    fn some_system(world: &mut World) {
        let mut param: PtrMut = world.into();
        immutable_reference(&param);
        mutable_reference(&mut param);
        _ = mutable_reference_return(&mut param);
        _ = mutable_reference_bounded_return(&mut param);
        _ = mutable_reference_bounded_return_complex(&mut param);
    }

    App::new().add_systems(Update, some_system).run();
}
