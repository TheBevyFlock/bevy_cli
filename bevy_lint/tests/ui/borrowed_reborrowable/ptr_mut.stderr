error: parameter takes `&mut PtrMut` instead of a re-borrowed `PtrMut`
  --> tests/ui/borrowed_reborrowable/ptr_mut.rs:17:22
   |
17 | fn mutable_reference(_param: &mut PtrMut) {
   |                      ^^^^^^^^^^^^^^^^^^^ help: use `PtrMut` instead: `mut _param: bevy::bevy_ptr::PtrMut<'_>`
   |
note: the lint level is defined here
  --> tests/ui/borrowed_reborrowable/ptr_mut.rs:5:9
   |
5  | #![deny(bevy::borrowed_reborrowable)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: parameter takes `&mut PtrMut` instead of a re-borrowed `PtrMut`
  --> tests/ui/borrowed_reborrowable/ptr_mut.rs:23:33
   |
23 | fn mutable_reference_return<'a>(_param: &'a mut PtrMut) -> usize {
   |                                 ^^^^^^^^^^^^^^^^^^^^^^ help: use `PtrMut` instead: `mut _param: bevy::bevy_ptr::PtrMut<'_>`

error: aborting due to 2 previous errors

